cyten/models/couplings.py

```diff
@@ -0,0 +1,757 @@
+                                  name: str = 'long-range quad coupling') -> Coupling:
+    r"""Long-range quadratic coupling (hopping and pairing) for bosons or fermions.
+
+    Quadratic couplings generally take the form
+    ``\sum_k -1 * t[k] a_i^\dagger(k) a_j(k) + delta[k] a_i^\dagger(k) a_j^\dagger(k) + h.c.``,
+    where the first term corresponds to hopping processes and the second term to superconducting
+    pairings. Here `a_i^\dagger(k)` and `a_j(k)` denote the creation and annihilation operators on
+    sites `i` and `j`, respectively. Sites `i` and `j` are assumed to be the first and the last
+    entry in `sites`, respectively. `k` denotes the boson / fermion species. Note the convention
+    for the order of the operators (with ``j > i``). This is in particular important for fermions
+    due to their anticommutation relations.
+
+    Parameters
+    ----------
+    t : float | list[float | None]
+        Hopping amplitudes. Add the corresponding hopping
+        ``\sum_k -1 * t[k] a_i^\dagger(k) a_j(k) + h.c.`` with the value as prefactor, where `k`
+        refers to the boson or fermion species. If there are multiple species and `t` is a `float`,
+        it is applied as hopping amplitude to all particle species.
+        `None` entries correspond to no hopping processes.
+    delta : float | list[float | None]
+        Superconducting pairing amplitudes. Add the corresponding pairing
+        ``\sum_k delta[k] a_i^\dagger(k) a_j^\dagger(k) + h.c.`` with the value as prefactor, where
+        `k` refers to the boson or fermion species. If there are multiple species and `delta` is a
+        `float`, it is applied as pairing amplitude to all particle species.
+        `None` entries correspond to no pairing processes.
+    """
+    # TODO test that this builds what we expect
+    assert len(sites) > 1, 'distance between the sites must be positive'
+    sites_ij = [sites[0], sites[-1]]
+    assert sites_ij[0].num_species == sites_ij[1].num_species
+    if not isinstance(t, (list, np.ndarray)):
+        t = [t] * sites[0].num_species
+    else:
+        assert len(t) == sites[0].num_species
+    if not isinstance(delta, (list, np.ndarray)):
+        delta = [delta] * sites[0].num_species
+    else:
+        assert len(delta) == sites[0].num_species
+    h = 0
+    for k, (t_k, delta_k) in enumerate(zip(t, delta)):
+        if t_k is None and delta_k is None:
+            continue
+        a_idk = sites_ij[0].creators[:, :, k]
+        a_jk = sites_ij[1].annihilators[:, :, k]
+        a_jdk = sites_ij[1].creators[:, :, k]
+        for n, site in enumerate(sites):
+            if isinstance(site, FermionicDOF):
+                if n == 0:
+                    jw_n = np.tensordot(site.creators[:, :, k:], site.annihilators[:, :, k:],
+                                        axes=[[2, 1], [2, 0]])
+                elif n == len(sites) - 1:
+                    jw_n = np.tensordot(site.creators[:, :, :k], site.annihilators[:, :, :k],
```

Comment:
again, this should simplify if we remove the sites where the operator doesnt act